### ISSUE TITLE
Null results from ExtractConnectedComponent

### ISSUE THREAD

**Issue by xiangtaoxu** [CloudCompare/CloudComPy#116]
Hi,

I found that when the actual number of components is larger than the `maxNumberComponents` parameter, `cc.ExtractConnectedComponents` will return 0 instead of truncate the connected components (like in the GUI).

I don't have a minimal reproducing example. I had this bug when processing a rather large point cloud. Here are the two screen shots for results that should have 120 connected components

1. 0 components
![image](https://github.com/CloudCompare/CloudComPy/assets/2780591/a465ea4b-8f43-4a19-bc9c-0db3cb10cbc7)

2. correct number of components with increased maxNumberComponents
![image](https://github.com/CloudCompare/CloudComPy/assets/2780591/c38c401d-d724-4b59-b27a-11327fe18313)

**Comment by prascle** [comment]
Hi, 
Issue reproduced, it will be fixed in the next release.

Regards,
Paul

**Comment by xiangtaoxu** [comment]
I am having some more issues with ExtractConnectedComponent. The function is causing some random segmentation fault during my application. The error seems to be nasty since I can't reproduce it every time. It usually happens when I try to extract lots of components....

I turned on cc_trace and below are the last few lines before one crash. Any ideas on the cause?
```
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [454] : ExtractConnectedComponents_py
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [481] : cloud
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [509] : ---
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [517] : ---
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [527] : ---
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [539] : total components: 1522 with 17 components of size > 1000
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [329] : createComponentsClouds_ 0
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [442] : [CreateComponentsClouds] 17 component(s) were created from cloud sub.clone
/home/paul/projets/CloudComPy/CloudComPy/pybind11/cloudComPy.cpp [572] : nbCloudDone: 1
```

**Comment by prascle** [comment]
This may not be easy without the debugger.

I took a look at the code, and found a bug related to the residual components (issue #95), which are incorrectly selected (the selected points are not the residual points). Anyway, I don't think this is the explanation for your bug.

After the last line of the trace, there is just the pybind11 C++ instruction that creates the return tuple (int, list, list).
Does it abort before return to Python ?
what happens if you add prints just after the return ?
```
res = cc.ExtractConnectedComponents(...)
print(res[0])
print(len(res[1]))
print(len(res[2]))
```

If you happen to have a not too large dataset on which the bug often appears, I can try with the debugger...

Regards,
Paul

**Comment by xiangtaoxu** [comment]
I will try adding more print statements. The challenge is that the occurrence seems to be totally random... I just tried running the same codes that crashed yesterday, it did not have any errors...

Currently, I wrote my own connected component extraction in python using scipy.spatial.kdtree so that our workflow can run. I will update the issue if we have more information

**Comment by xiangtaoxu** [comment]
Update: ExtractConnectedComponent has been working fine in the last couple of weeks. Still don't know what caused the error before. But it seems to be OK!

Meanwhile, we used the functions a lot in our application and it can be slow when the components are over 10,000. I wonder whether it is easy to adapt the algorithm to use multi-processing/multi-threading? I noticed the function only uses one-core for now.
